### PR TITLE
Element tracking

### PR DESCRIPTION
**Pull request by robin-marr-ft**
/cc @AlbertoElias @kavanagh

### CODE REVIEW

File: test/viewport.test.js [hunk 1]
```diff
@@ -54,4 +84,19 @@ describe('o-viewport', function() {
...
+
+	it('should stop listening to scroll event', function(done) {
+		oViewport.stopListeningTo('scroll');
+		document.body.addEventListener('oViewport.scroll', function(ev) {
+			expect(false).to.be.true;
+			done();
+		});
+
+		setTimeout(function () {
+			expect(true).to.be.true;
```
Comment (kaelig) [inline test/viewport.test.js]: That's a funny one

File: test/viewport.test.js [hunk 2]
```diff
@@ -54,4 +84,19 @@ describe('o-viewport', function() {
 	it('should get the orientation of the viewport', function() {
 		expect(oViewport.getOrientation() === 'portrait' || oViewport.getOrientation() === 'landscape').to.be(true);
 	});
+
+	it('should stop listening to scroll event', function(done) {
+		oViewport.stopListeningTo('scroll');
+		document.body.addEventListener('oViewport.scroll', function(ev) {
+			expect(false).to.be.true;
```
Comment (kaelig) [inline test/viewport.test.js]: quote-william-s -burroughs-nothing-is-true-everything-is-permitted-124339_1

Comment (robin-marr-ft) [inline test/viewport.test.js]: Forgot I was supposed to look up how to do that properly, I've updated it
now.


On 24 March 2015 at 15:58, Kaelig <notifications@github.com> wrote:

> In test/viewport.test.js
> <https://github.com/Financial-Times/o-viewport/pull/17#discussion_r27042323>
> :
>
> > @@ -54,4 +84,19 @@ describe('o-viewport', function() {
> >  	it('should get the orientation of the viewport', function() {
> >  		expect(oViewport.getOrientation() === 'portrait' || oViewport.getOrientation() === 'landscape').to.be(true);
> >  	});
> > +
> > +	it('should stop listening to scroll event', function(done) {
> > +		oViewport.stopListeningTo('scroll');
> > +		document.body.addEventListener('oViewport.scroll', function(ev) {
> > +			expect(false).to.be.true;
>
> [image: quote-william-s
> -burroughs-nothing-is-true-everything-is-permitted-124339_1]
> <https://cloud.githubusercontent.com/assets/85783/6806103/aca69098-d23e-11e4-8c7e-81a4c5560005.png>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Financial-Times/o-viewport/pull/17/files#r27042323>.
>

-- 

------------------------------
This email was sent by a company owned by Pearson plc, registered office at 
80 Strand, London WC2R 0RL.  Registered in England and Wales with company 
number 53723.

File: .gitignore [hunk 3]
```diff
@@ -2,5 +2,7 @@
 .sass-cache/
 bower_components/
 node_modules/
+demos/
```
Comment (AlbertoElias) [inline .gitignore]: Why are you ignoring demos? You should just be ignoring demos/local

**Comment by commuterjoy** [comment]
out of interest, what's this for?

File: .gitignore (hunk 3 above)
Comment (robin-marr-ft) [inline .gitignore]: Because when I run obt demo --local I get the below tree, this gitignore
config will ignore everything in red:

demos
├── element-tracking.html
├── local
│   ├── element-tracking.css
│   ├── element-tracking.html
│   └── element-tracking.js
└── src

    ├── config.json

    ├── element-tracking.js

    ├── element-tracking.mustache

    └── element-tracking.scss

On 25 March 2015 at 10:46, Alberto Elias <notifications@github.com> wrote:

> In .gitignore
> <https://github.com/Financial-Times/o-viewport/pull/17#discussion_r27109867>
> :
>
> > @@ -2,5 +2,7 @@
> >  .sass-cache/
> >  bower_components/
> >  node_modules/
> > +demos/
>
> Why are you ignoring demos? You should just be ignoring demos/local
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Financial-Times/o-viewport/pull/17/files#r27109867>.
>

-- 

------------------------------
This email was sent by a company owned by Pearson plc, registered office at 
80 Strand, London WC2R 0RL.  Registered in England and Wales with company 
number 53723.

**Comment by robin-marr-ft** [comment]
Hi Matt,

We found Paul and I were duplicating effort, o-ads needs to track elements
for lazy loading, sticky and refresh in view so we decided to move the
functionality into an origami module so it can be shared, after discussing
with the origami team we decided to add it to o-viewport

Cheers
Rob

On 25 March 2015 at 10:52, Matt Chadburn <notifications@github.com> wrote:

> out of interest, what's this for?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Financial-Times/o-viewport/pull/17#issuecomment-85981170>
> .
>

-- 

------------------------------
This email was sent by a company owned by Pearson plc, registered office at 
80 Strand, London WC2R 0RL.  Registered in England and Wales with company 
number 53723.

File: .gitignore (hunk 3 above)
Comment (AlbertoElias) [inline .gitignore]: You're supposed to get everything that's not in local

Comment (robin-marr-ft) [inline .gitignore]: I see, sorry, I was told not to commit the files in root, I've fixed this
now.

On 25 March 2015 at 12:15, Alberto Elias <notifications@github.com> wrote:

> In .gitignore
> <https://github.com/Financial-Times/o-viewport/pull/17#discussion_r27114961>
> :
>
> > @@ -2,5 +2,7 @@
> >  .sass-cache/
> >  bower_components/
> >  node_modules/
> > +demos/
>
> You're supposed to get everything that's not in local
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Financial-Times/o-viewport/pull/17/files#r27114961>.
>

-- 

------------------------------
This email was sent by a company owned by Pearson plc, registered office at 
80 Strand, London WC2R 0RL.  Registered in England and Wales with company 
number 53723.

**Comment by wheresrhys** [comment]
My gut feeling is that this should have oViewport as a dependency and be its own module. It seems like a massive departure from the original intention of the module, which was a minimal utility to discourage performance bottlenecks and normalise a few bowser quirks.

The code looks good, but I really don't think this is the place for it

**Comment by kavanagh** [comment]
Agree with @wheresrhys.

**Comment by commuterjoy** [comment]
I talked with @samgiles and @AlbertoElias about a having a shared instrumentation that each origami consumer can use to measure usage/behaviour etc. Eg, o-viewability would be a lot clearer in scope.

**Comment by AlbertoElias** [comment]
Ok, let's discuss again at the meeting this week. It was discussed last week and agreed to do it like this, but if @robin-marr-ft is ok with it, let's talk about having a separate module just for this, or for this and maybe other useful features

File: main.js [hunk 4]
```diff
@@ -1,129 +1,69 @@
...
+			console.error(err);
+		}
 		return;
 	}
 
-	window.addEventListener('resize', debounce(function(ev) {
-		broadcast('resize', {
-			viewport: getSize(),
-			originalEvent: ev
+	if (elements.length) {
```
Comment (AlbertoElias) [inline main.js]: Why is this necessary?

File: README.md [hunk 5]
```diff
@@ -1,34 +1,48 @@
 # o-viewport [![Build Status](https://travis-ci.org/Financial-Times/o-viewport.svg?branch=master)](https://travis-ci.org/Financial-Times/o-viewport)
 
-Utility for attaching debounced listeners to resize, scroll and orientation events on `window`.
+Utility for attaching debounced listeners to resize, scroll, orientation and visibility events on `window`.
 
-*Note: within the module's API and in the documentation below `orientation` is used instead of `orientationchange`, but the actual browser event listened to is `orientationchange`*
+*Note: within the module's API and in the documentation below `orientation` and `visibility` are used instead of `orientationchange` or `visibilitychange`, but the actual browser event listened to is `orientationchange` or `visibilitychange`*
```
Comment (AlbertoElias) [inline README.md]: should be and instead of or?

File: README.md [hunk 6]
```diff
@@ -1,34 +1,48 @@
...
 
 ## Methods
 
 ### `o-viewport#listenTo(eventType)`
-Attaches a debounced/throttled (as appropriate) listener to events on window [`resize`, `scroll` or `orientation`] which in turn fires events within the `oViewport` namespace (see **Events** below).
+Attaches a debounced/throttled (as appropriate) listener to events on window [`resize`, `scroll`, `orientation` or `visibility`] which in turn fires events within the `oViewport` namespace (see **Events** below).
+
+### `o-viewport#trackElements(selector)`
+Tracks the selected elements and report the percentage of the element currently on the screen, changes to the percentage in view fires events within the `oViewport` namespace (see **Events** below).
```
Comment (AlbertoElias) [inline README.md]: reports

File: test/viewport.test.js [hunk 7]
```diff
@@ -1,11 +1,30 @@
...
 
+function isPhantom(){
+	return /PhantomJS/.test(navigator.userAgent);
+}
+
 describe('o-viewport', function() {
+	before(function(){
+		if(isPhantom()) {
+			// hack to make test run in PhantomJS
+			// it doesn't support visibilitychange
```
Comment (AlbertoElias) [inline test/viewport.test.js]: Maybe add support to it to the polyfill service and document it in the origami.josn

File: README.md [hunk 8]
```diff
@@ -1,34 +1,48 @@
...
+
+If the selector argument is omitted `[data-o-viewport-track]` is used.
```
Comment (AlbertoElias) [inline README.md]: comma after omitted

File: README.md [hunk 9]
```diff
@@ -1,34 +1,48 @@
...
+
+*Note: Enabling element tracking will also attach all other o-viewport events.*
```
Comment (AlbertoElias) [inline README.md]: Why?

File: README.md [hunk 10]
```diff
@@ -1,34 +1,48 @@
...
-*Note: within the module's API and in the documentation below `orientation` is used instead of `orientationchange`, but the actual browser event listened to is `orientationchange`*
+*Note: within the module's API and in the documentation below `orientation` and `visibility` are used instead of `orientationchange` or `visibilitychange`, but the actual browser event listened to is `orientationchange` or `visibilitychange`*
 
 ## Methods
 
 ### `o-viewport#listenTo(eventType)`
-Attaches a debounced/throttled (as appropriate) listener to events on window [`resize`, `scroll` or `orientation`] which in turn fires events within the `oViewport` namespace (see **Events** below).
+Attaches a debounced/throttled (as appropriate) listener to events on window [`resize`, `scroll`, `orientation` or `visibility`] which in turn fires events within the `oViewport` namespace (see **Events** below).
+
+### `o-viewport#trackElements(selector)`
```
Comment (AlbertoElias) [inline README.md]: selectorS? Is it an array?

File: README.md [hunk 11]
```diff
@@ -1,34 +1,48 @@
...
 Provides a reasonably reliable way (more so than `window.orientation`) of obtaining the current orientation of the viewport.
 
+### `o-viewport#getVisibility()`
+Provides a reasonably reliable way of obtaining the current visibilty of the viewport.
+
 ### `o-viewport#getSize()`
-Provides a reliable way of obtaining the current dimensions of the viewport. returns an object with the properties `width` and `height`
+Provides a reliable way of obtaining the current dimensions of the browser window. returns an object with the properties `width` and `height`
+
+### `o-viewport#getViewportSize()`
```
Comment (AlbertoElias) [inline README.md]: Is this a breaking change?

File: main.js [hunk 12]
```diff
@@ -1,129 +1,69 @@
...
-	return false;
 }
 
-function listenToResize() {
-
-	if (init('resize')) {
+function trackElements(selector){
+	var elements;
+	selector = typeof selector === 'string' ? selector : '[data-o-viewport-track]';
+	try{
```
Comment (AlbertoElias) [inline main.js]: Is this necessary? If it can't find anything, it will return an empty array

File: src/tracking.js [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+				inView: !!inView,
+				inViewPercentage: inView
+			}, element.node);
+		}
+	});
+}
+
+module.exports = {
+	track: Element,
+	tracked: tracked,
```
Comment (AlbertoElias) [inline src/tracking.js]: would call it trackedElements. tracked I thought would be a true-false value

File: src/tracking.js [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+	tracked.forEach(function (element){
+		element.getNodePosition();
+		update();
+	});
+}
+
+function update(){
+	tracked.forEach(function (element){
+		if(element.lastResult !== element.percentInViewport()){
+			var inView = element.lastResult = element.percentInViewport();
```
Comment (AlbertoElias) [inline src/tracking.js]: why set a new variable instead of just .lastResult?

**Comment by AlbertoElias** [comment]
Should we close this in favour of https://github.com/Financial-Times/o-element-visibility

**Comment by wheresrhys** [comment]
:+1:


On 21 May 2015 at 14:18, Alberto Elias <notifications@github.com> wrote:

> Should we close this in favour of
> https://github.com/Financial-Times/o-element-visibility
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Financial-Times/o-viewport/pull/17#issuecomment-104275030>
> .
>

-- 

------------------------------
This email was sent by a company owned by Pearson plc, registered office at 
80 Strand, London WC2R 0RL.  Registered in England and Wales with company 
number 53723.

**Comment by robin-marr-ft** [comment]
I've cleaned this up now. I've added a couple of things used by element visibility, namely the visibility event and the getScrollPosition function. The main file was getting a bit long so spilt some methods out into a utils file. Everybody happy if I merge?

**Comment by wheresrhys** [comment]
Looks good to me (though it's not my decision)

**Comment by AlbertoElias** [comment]
Are things here that it's only use case is in o-element-viewability or that you think should maybe be in there directly because you won't use them if you're not using that module?

**Comment by robin-marr-ft** [comment]
Nope, all that stuff has been removed.

**Comment by AlbertoElias** [comment]
lgtm then
